### DISCUSSION TITLE
Self-hosted media library integration

### DISCUSSION THREAD

**Discussion by theetrain** [decaporg/decap-cms#5982] [Ideas]
Hi, I've successfully integrated NetlifyCMS with Cloudinary and things went smoothly. Cloudinary provides its own [upload widget](https://cloudinary.com/documentation/upload_widget) that integrates nicely with NetlifyCMS, allowing me to log into my Cloudinary account and upload and insert media.

I'm wondering if there's an appetite for similar integration with self-hosted media, such as with [Nextcloud](https://nextcloud.com/) or [Owncloud](https://owncloud.com/). I assume what's needed is for one of these services to also have an 'upload widget' that could accept new file uploads and return URL strings to 'insert' into its respective page being edited. 

Has anyone achieved results for this already and is willing to share? I also reached out to Nextcloud to see if something can be done: https://help.nextcloud.com/t/netlifycms-integration/127266

**Comment by erezrokah** [comment]
> I assume what's needed is for one of these services to also have an 'upload widget' that could accept new file uploads and return URL strings to 'insert' into its respective page being edited.

🆙 This. The widget does most of the heavy lifting. The CMS integration shouldn't have a lot of code to it. See https://github.com/netlify/netlify-cms/blob/0191f25c51ef66292c920a93078697643c8dd68b/packages/netlify-cms-media-library-cloudinary/src/index.js

**Comment by privatemaker** [comment]
I found the following open-source alternatives which need further investigation:

- Piwigo - https://github.com/Piwigo
- Thumbor - https://github.com/thumbor/thumbor
- Imageflow - https://github.com/imazen/imageflow
- Glide - https://github.com/thephpleague/glide
- ZeStream - https://github.com/zestream
- Tachyon - https://github.com/humanmade/tachyon (AWS)

I'm not sure which of these are most viable, but I will try to investigate more properly soon.

If there's any devs who have poked at the media adapter widget code and would like to comment on what is ideal / bad, that would be helpful 😸
